8323698: Class use page does not include extends/implements type arguments

liach opened the pull request:
Currently in the javadoc tool, the generated class use page does not track the occurrences of a class in the type arguments of the extends or implements list, where they can appear. (See more details on the JBS issue) For example:
```java
public class One {}
```
```java
import java.util.*;
public class Two extends ArrayList<One> implements Comparator<One> {
}
```

This patch proposes to add new 2 usage categories, "Classes that extend types with arguments of type" and "Classes that implement types with arguments of type" to describe the missing usage scenarios.

A preview of the new use page with the 2 new categories, generated based on the example above, can be found here: https://cr.openjdk.org/~liach/8323698-javadocuse/doc/class-use/One.html

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323698](https://bugs.openjdk.org/browse/JDK-8323698): Class use page does not include extends/implements type arguments (**Bug** - P4)


### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17414/head:pull/17414` \
`$ git checkout pull/17414`

Update a local copy of the PR: \
`$ git checkout pull/17414` \
`$ git pull https://git.openjdk.org/jdk.git pull/17414/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17414`

View PR using the GUI difftool: \
`$ git pr show -t 17414`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17414.diff">https://git.openjdk.org/jdk/pull/17414.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17414#issuecomment-1890812117)

hns commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:
```diff
@@ -380,6 +380,8 @@ doclet.ClassUse_Classes.in.0.used.by.1=Classes in {0} used by {1}
 doclet.ClassUse_Annotation=Classes in {1} with annotations of type {0}
 doclet.ClassUse_TypeParameter=Classes in {1} with type parameters of type {0}
+doclet.ClassUse_SuperclassTypeParameter=Classes in {1} that extend types with arguments of type {0}
+doclet.ClassUse_InterfaceTypeParameter=Classes in {1} that implement types with arguments of type {0}
```
Not sure about the wording here: I think it should be "type arguments" instead of just "arguments", and maybe "extend classes" and "implement interfaces"?

liach replied:
Done. Also pulled latest master, the preview on CR is updated too.

jonathan-gibbons commented on test/langtools/jdk/javadoc/doclet/testUseOption/TestUseOption.java:
```diff
@@ -191,4 +195,61 @@ public void test3() {
+
+    @Test
+    public void testSuperclassAndInterfaceTypeArgument(Path base) throws IOException {
+        Files.createDirectories(base);
+        Path out = base.resolve("out");
+        Path src = base.resolve("src");
+
+        Files.createDirectories(out);
+        Files.createDirectories(src);
+        Path one = src.resolve("One.java");
+        Files.writeString(one, """
+            public class One {}
+            """);
+        Path two = src.resolve("Two.java");
+        Files.writeString(two, """
+            import java.util.*;
+            public class Two extends ArrayList<One> implements Comparator<One> {
+            }
+            """);
```
This is not wrong, but if you use `ToolBox.writeJavaFiles` the amount of boilerplate is reduced.

jonathan-gibbons commented on test/langtools/jdk/javadoc/doclet/testUseOption/TestUseOption.java:
```diff
@@ -191,4 +195,61 @@ public void test3() {
+        );
+        checkExit(Exit.OK);
+
+        checkOrder("class-use/One.html", """
```
This method call would be easier to read if the text blocks were indented relative to the opening `"""`.  See the other examples earlier in this file and follow the style of those examples.

liach commented:
@hns @jonathan-gibbons Sorry for pinging, can you review this pr again?

pavelrappo reviewed (commented):
It does not feel generic enough. (No pun intended.) What about using a class or interface as a bound as in this example?

    interface MyList<T extends One & Two<Three>> extends java.util.List<T>

pavelrappo commented:
> It does not feel generic enough. (No pun intended.) What about using a class or interface as a bound as in this example?
> 
> ```
> interface MyList<T extends One & Two<Three>> extends java.util.List<T>
> ```

I checked, and your patch seems to automatically work for the case that I suggested. Good! However, please take into account the following when refining label text (standard.properties): 

  * interfaces can also be generic,
  * an interface extends other interfaces, not implements them,
  * there's probably a difference between a class or interface being used as a type argument (i.e. `MyList<One>`) and it being used as a bound (i.e. `MyList<T extends One>`).

liach commented:
Thanks for your review, Pavel! I've addressed your changes and updated the preview on CR.

1. `interface MyList<T extends One & Two<Three>> extends java.util.List<T>`
This case is already perfectly handled by existing `doclet.ClassUse_TypeParameter`, since `One` `Two` `Three` here exists in a type parameter's bounds, which will be scanned by a visitor and recorded correctly.
2. For generic interfaces, I have reworded the usage categories; closer to the simple subclass version.
3. Have split out the subinterfaces in a separate category and added it to tests.
4. These being used as a type argument and as a bound are unfortunately not distinguished by our existing type parameter handling, in both type parameter declarations in classes/methods and parameterized type usages in field/method types. I don't think it's in this patch's scope to change this behavior, and if this should be changed, all those sites should be updated as well.

Feel free to suggest improvments to the wordings!

liach commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:
```diff
@@ -280,7 +280,8 @@ doclet.help.use.body=\
     Each documented package, class and interface has its own Use page.  This page describes what \
     packages, classes, methods, constructors and fields use any part of the given class or \
     package. Given a class or interface A, its Use page includes subclasses of A, fields declared \
-    as A, methods that return A, and methods and constructors with parameters of type A.  \
+    as A, methods that return A, methods and constructors with parameters of type A, and \
+    subclasses with parameters of type A.  \
     You can access this page by first going to the package, class or interface, then clicking on \
     the USE link in the navigation bar.
```
This is an improvement. Applying as I don't see any problem in your wording. ("class or interface" is an established term)

liach commented:
I will wait 24 hours before integration to allow for more reviews.

hns reviewed (approved):
Looks good to me.

liach commented:
/integrate
